### ISSUE TITLE
ImfAttribute symbol visibility not quite right

### ISSUE THREAD

**Issue by meshula** [AcademySoftwareFoundation/openexr#982]
Here is an excerpt illustrating the issue.

Here, Imf_3_0::Attribute has inconsistent visibility versus the derived class.

```
[ 33%] Building CXX object src/lib/OpenEXR/CMakeFiles/OpenEXR.dir/ImfStringVectorAttribute.cpp.obj
ImfStringVectorAttribute.cpp
C:\Projects\Lucasfilm\openexr\src\lib\OpenEXR\ImfStringVectorAttribute.cpp(80): warning C4275: non dll-interface class 'Imf_3_0::Attribute' used as base for dll-interface class 'Imf_3_0::TypedAttribute<Imf_3_0::StringVector>'
C:\Projects\Lucasfilm\openexr\src\lib\OpenEXR\ImfAttribute.h(28): note: see declaration of 'Imf_3_0::Attribute'
C:\Projects\Lucasfilm\openexr\src\lib\OpenEXR\ImfStringVectorAttribute.h(29): note: see declaration of 'Imf_3_0::TypedAttribute<Imf_3_0::StringVector>'
```

Here we are attempting to export something that can't be exported.

```
C:\Projects\Lucasfilm\openexr\src\lib\OpenEXR\ImfStringVectorAttribute.cpp(80): warning C4251: 'Imf_3_0::TypedAttribute<Imf_3_0::StringVector>::_value': class 'std::vector<std::string,std::allocator<std::string>>' needs to have dll-interface to be used by clients of class 'Imf_3_0::TypedAttribute<Imf_3_0::StringVector>'
C:\Projects\Lucasfilm\openexr\src\lib\OpenEXR\ImfStringVectorAttribute.h(28): note: see declaration of 'std::vector<std::string,std::allocator<std::string>>'
```

**Comment by kdt3rd** [comment]
@meshula - can you provide a bit more detail? I presume this is under MSVC, but what version? I do not see this when compiling with mingw, and didn't see any warnings of the sort in the msvc builds as part of the CI, but I may have missed them.

**Comment by meshula** [comment]
The CI is weirdly clean. I just did a clean pull to verify.

````
git@github.com:openexr/openexr.git
mkdir build
cd build
cmake ../openexr -G "Visual Studio 16 2019" -Ax64 -DCMAKE_INSTALL_PREFIX=./install
cmake --build . --target install --config Release
````

There is an enormous amount of warning spew compared to what we see in the CI, including the dll-interface warning. Here is the tiniest of excerpts as confirmation. This one repeats a hundred times at least. There's narrowing warnings, and deprecation warnings, and the old favorite, CRT_SECURE warnings. My scrollback buffer is 2000 lines long, and it quickly overflows...

```
C:\Projects\Lucasfilm\openexr\src\lib\OpenEXR\ImfAttribute.h(28): message : see declaration of 'Imf_3_0::Attribute' [C:\Projects\Lucasfilm\build\src\lib\OpenEXR\OpenEXR.vcxproj]
C:\Projects\Lucasfilm\openexr\src\lib\OpenEXR\ImfChannelListAttribute.h(25): message : see declaration of 'Imf_3_0::TypedAttribute<Im
f_3_0::ChannelList>' [C:\Projects\Lucasfilm\build\src\lib\OpenEXR\OpenEXR.vcxproj]
C:\Projects\Lucasfilm\openexr\src\lib\OpenEXR\ImfChannelListAttribute.cpp(136,104): warning C4251: 'Imf_3_0::TypedAttribute<Imf_3_0::
ChannelList>::_value': class 'Imf_3_0::ChannelList' needs to have dll-interface to be used by clients of class 'Imf_3_0::TypedAttribu
te<Imf_3_0::ChannelList>' [C:\Projects\Lucasfilm\build\src\lib\OpenEXR\OpenEXR.vcxproj]
C:\Projects\Lucasfilm\openexr\src\lib\OpenEXR\ImfChannelList.h(89): message : see declaration of 'Imf_3_0::ChannelList' [C:\Projects\
Lucasfilm\build\src\lib\OpenEXR\OpenEXR.vcxproj]
  ImfFloatAttribute.cpp
C:\Projects\Lucasfilm\openexr\src\lib\OpenEXR\ImfFloatAttribute.cpp(28,66): warning C4275: non dll-interface class 'Imf_3_0::Attribut
e' used as base for dll-interface class 'Imf_3_0::TypedAttribute<float>' [C:\Projects\Lucasfilm\build\src\lib\OpenEXR\OpenEXR.vcxproj
]
C:\Projects\Lucasfilm\openexr\src\lib\OpenEXR\ImfAttribute.h(28): message : see declaration of 'Imf_3_0::Attribute' [C:\Projects\Luca
sfilm\build\src\lib\OpenEXR\OpenEXR.vcxproj]
C:\Projects\Lucasfilm\openexr\src\lib\OpenEXR\ImfFloatAttribute.h(25): message : see declaration of 'Imf_3_0::TypedAttribute<float>'
[C:\Projects\Lucasfilm\build\src\lib\OpenEXR\OpenEXR.vcxproj]
  ImfFrameBuffer.cpp
````

**Comment by meshula** [comment]
[
[spew.txt](https://github.com/AcademySoftwareFoundation/openexr/files/6176681/spew.txt)
](url)

I've attached spew.txt so you can see it in its full glory. It's 4,500 lines of warnings.

**Comment by cary-ilm** [comment]
@meshula, were you able to determine anything further regarding the Imf_3_0::Attribute symbol visibility?

**Comment by meshula** [comment]
I am investigating whether there's a build setting issue with things (cmake setting? preprocessor definitions?) not propagating from OpenEXR to Imath, as a mismatch could be the root cause, but haven't got a result yet.

**Comment by ignus2** [comment]
I can confirm, the amount of warnings went through the roof with MSVC builds.
As I see it is not an easy setting fix. The core of the issue for some of them is that the class "Attribute" is not exported under MSVC (IMF_EXPORT_TYPE is defined as empty in ImfExport.h), whereas the classes deriving from "Attribute" are.
For example in ImfFloatAttribute.cpp:
```
template class IMF_EXPORT_TEMPLATE_INSTANCE TypedAttribute<float>;
```
where IMF_EXPORT_TEMPLATE_INSTANCE ends up being declspec(dllexport) under MSVC, hence the warning. In fact, this is exactly what the warning is telling us (copied from above):
```
ImfFloatAttribute.cpp(28,66): warning C4275: non dll-interface class 'Imf_3_0::Attribute' used as base for dll-interface class 'Imf_3_0::TypedAttribute<float>'
```
As I see most of C4275 is caused by the class "Attribute".

The other DLL related warning is C4251. I have 482 warnings of C4275 and 386 warnings of C4251, each having additional "see declaration ..." lines in the warning log, so those two take up a significant portion of the warning log.

I just noticed #983 which summarizes the issues as well.

**Comment by meshula** [comment]
Update: @ignus2 @kdt3rd It runs out the macro set ups are all correct as is, in terms of making OpenEXR and Imath functional as shared libraries under MSVC. Attempting to resolve the reported warnings by class exporting, through the IMF_EXPORT_TYPE macro leads to a new issue, which is that the classes Attribute is templated on, also need to be class exported. However, in ImathVec, we delete the specializations on normalize for integer, because they are non-sensical. If we class export the Vec template, the compiler generates an error, not a warning, because class exporting a template force instantiates all the members. Later in the file we delete them, and the compiler does not allow that to override the class export, and compilation fails. VecN<int>::normalize is not the only case of a problem, but it's illustrative.

Unless anyone has a clever idea, I think they only reasonable resolution is to suppress the 4275 and 4251 warnings, as I have proposed in https://github.com/AcademySoftwareFoundation/openexr/pull/983.

**Comment by cary-ilm** [comment]
#983 is currently a draft, can you turn it into a reviewable/mergeable PR? If we want this for the 3.0 release it needs to go in asap.

**Comment by ignus2** [comment]
@meshula The main question I'd have is why are `TypedAttribute<float>` etc. specializations class-exported? That seems to be the cause of the warnings primarily.

**Comment by meshula** [comment]
The rules are vague and hard to parse~ https://docs.microsoft.com/en-us/cpp/cpp/general-rules-and-limitations?view=msvc-160 ~ 

The templates are class exported for typeid comparisons under gcc/clang, doesn't necessarily mean they need to also class exported under msvc. We've tried a number of permutations of macros to satisfy for all platforms, including attempting not class exporting on msvc. This is the only combo so far, that works on all platforms. If there's a permutation that is warning-clean on windows, and macros can be platform-jiggered to work that would be great ~ we just haven't found that combo by reading or experimentation yet~! 

Definitely open to a PR, if you've got a solution. I'm sure the solution once found will be head smackingly obvious in retrospect, but, currently stumped :)

**Comment by cary-ilm** [comment]
Culling old issues. I think the symbol visibility issues have been resolved, and the warnings almost quashed as well.
